protob/messages-eos.proto

```diff
@@ -0,0 +1,265 @@
+    optional string wif_public_key = 1;     // EOS pub key in Base58 encoding
+    optional bytes raw_public_key = 2;      // Raw public key
+}
+ /**
+ * Request: Ask device to sign transaction
+ * @start
+ * @next EosTxRequest
+ * @next Failure
+ */
+ message EosSignTx {
+    repeated uint32 address_n = 1;                // BIP-32 path to derive the key from master node 44'/194'/0'
+    optional bytes chain_id = 2;                  // 256-bit long chain id
+    optional EosTxHeader header = 3;              // EOS transaction header
+    optional uint32 num_actions = 4;              // number of actions 
+     /**
+    * Structure representing EOS transaction header
+    */
+    message EosTxHeader {
+        required uint32 expiration = 1;           // time at which transaction expires
+        required uint32 ref_block_num = 2;        // 16-bit specifies a block num in the last 2^16 blocks.
+        required uint32 ref_block_prefix = 3;     // specifies the lower 32 bits of the blockid at get_ref_blocknum
+        required uint32 max_net_usage_words = 4;  // upper limit on total network bandwidth (in 8 byte words) billed for this transaction
+        required uint32 max_cpu_usage_ms = 5;     // 8-bit upper limit on the total CPU time billed for this transaction
+        required uint32 delay_sec = 6;            // number of seconds to delay this transaction for during which it may be canceled.
+    }   
+}
+ /**
+ * Response: Device asks to upload next action
+ * @next EosTxActionAck
+ */
+message EosTxActionRequest {
+    optional uint32 data_size = 1;
+}
+ /**
+ * Request: Next action data that needs to be uploaded
+ * @next EosTxActionRequest
+ * @next EosSignedTx
+ * @next Failure
+ */
+message EosTxActionAck {
+    optional EosActionCommon common = 1;
+    optional EosActionTransfer transfer = 2;
+    optional EosActionDelegate delegate = 3;
+    optional EosActionUndelegate undelegate = 4;
+    optional EosActionRefund refund = 5;
+    optional EosActionBuyRam buy_ram = 6;
+    optional EosActionBuyRamBytes buy_ram_bytes = 7;
+    optional EosActionSellRam sell_ram = 8;
+    optional EosActionVoteProducer vote_producer = 9;     
+    optional EosActionUpdateAuth update_auth = 10;   
+    optional EosActionDeleteAuth delete_auth = 11;
+    optional EosActionLinkAuth link_auth = 12;
+    optional EosActionUnlinkAuth unlink_auth = 13;
+    optional EosActionNewAccount new_account = 14;
+    optional EosActionUnknown unknown = 15;
+     /**
+    * Structure representing asset type
+    */
+    message EosAsset {
+        optional sint64 amount = 1;
+        optional uint64 symbol = 2;           // Lowest 8-bits used for precision.    
+    }
+    /**
+    * Structure representing action permission level
+    */
+    message EosPermissionLevel {
+        optional uint64 actor = 1;
+        optional uint64 permission = 2;
+    }
+
+    /**
+    * Structure representing auth key
+    */
+    message EosAuthorizationKey {
```

Comment:
If you add an optional `repeated uint32 address_n`, it allows for better UX. Instead of having to remember/write down pubkeys in cases where the device can compute it, in those cases you can show the BIP32 path instead.

Reply:
Can you describe the case when it will be applicable?

Reply:
Suppose you want to `UpdateAuth` and move your `@active` signing permission to a different HD path on your device. The device-side ux could be something like:

    Do you want to assign memememememe@active permission to EOS Account #1 @active?

assuming it was a valid slip48 path, or:

    Do you want to assign memememememe@active permission to m/42'/17'/32'/57?

if it was not. Contrast that with not providing an `address_n`, which forces the dialog to be:

    Do you want to assign memememememe@active permission to EOS62cPUiWnLqbUjiBMxbEU4pm4Hp5X3RGk4KMTadvZNygjX72yHW?

I've implemented that in our port here: https://github.com/keepkey/keepkey-firmware/blob/7a9db57cd4b15304b478443ea2b2ecdb46c51668/lib/firmware/eos-contracts/eosio.system.c#L507

Reply:
On the device side it looks handy. But It may bring inconvenience on client side. Signed Transaction body has to have key in plane format “EOS5kuQ...”, “PUB_K1_5kuQ...” or “PUB_R1_...” so you have to retrieve key from the device for a new path before generation the transaction body.

Reply:
address_n here is additional *optional* data, so it allows better ux in these case but doesn't force it. Clients are free to ignore it in our implementation, and the wif version of the pubkey would be shown in that case.

Reply:
In practice, it's very easy to handle client-side, since when transferring an account within the device, the UI is going to know what the path and pubkey are well before the device is asking for confirmation.... if you got to that point without this, you would have needed the pubkey anyway.